### DISCUSSION TITLE
Hosting in Virtual Directory (GitHub Pages)

### DISCUSSION THREAD

**Discussion by JoshClose** [statiqdev/Discussions#56] [Q&A]
I'll be deploying this by just outputting the build to the `/docs` folder of my repository. That means it will be hosted from `owner.githubpages.io/repository`. I'm trying to figure out the combination of settings to build it to work like that. I also would like to be able to run in a virtual directory during `preview` if possible.

I set
```yml
LinkRoot: /repository
```

All this seems to do is spit out errors when I run saying "Could not validate relative link" because my links don't start with `/repository`.

I see issue https://github.com/statiqdev/Statiq.Web/issues/735 but don't see a setting to enable that in Statiq Web.

It seems to work if I use the link shortcode. Is there a way to turn it on globally or globally or for markdown pages?

**Comment by daveaglick** [comment]
Bummer, looks like the `MarkdownPrependLinkRoot` setting didn't come over from the old blog recipe. I'll open an issue to add that back in the next version (https://github.com/statiqdev/Statiq.Web/issues/947).

In the meantime, you can modify the Markdown [template](https://statiq.dev/web/content-and-data/templates/) to enable prepending the link root since the actual `RenderMarkdown` module has support for it (there just isn't a setting in Statiq Web to turn it on):

```
bootstrapper
// ...
  .ModifyTemplate(MediaTypes.Markdown, module => ((RenderMarkdown)module).PrependLinkRoot(true))
```

As for previewing, you can have the preview server put the site at a virtual directory with `--virtual-dir <PATH>` on your command line. That combined with the `LinkRoot` setting will let you serve the site locally as if it were in the subfolder on GH Pages.

Let me know if this does the trick!

**Reply by JoshClose** [reply]
Thanks! I didn't see the `--virtual-dir <PATH>` arg here either https://statiq.dev/web/running/command-line.

**Reply by JoshClose** [reply]
Is there a way to have the `LinkRoot` injected into razor pages? I've been doing `~/scripts/header.js` which outputs `/scripts/header.js` but when using `LinkRoot` and `--virtual-dir`, it's not adding the `/virtual` to it. I don't think using `@Document.GetLink()` is correct, because I want it to be the site root, and not the location of the current document.

**Reply by JoshClose** [reply]
Is `Context.GetLink` the way I should do that?

**Reply by JoshClose** [reply]
That works, just asking in case there is a way to automatically have `~/` do it for me.

**Reply by JoshClose** [reply]
Is it normal to have every markdown link have a warning `Could not validate relative link` when using a virtual directory, or am I doing something wrong? If that's normal, I can just run outside a virtual while editing, then test and publish inside a virtual.

**Reply by JoshClose** [reply]
I'm also running into this issue in the markdown files now. https://github.com/statiqdev/Statiq.Web/issues/744 I don't see how the workaround will work with markdown.

**Reply by JoshClose** [reply]
I created a feature request for some link stuff which I'm willing to take a try at. https://github.com/statiqdev/Statiq.Framework/issues/170

**Reply by daveaglick** [reply]
> Is Context.GetLink the way I should do that?

Yep, the `GetLink()` methods are the right way. `~/` is generally a server-side convention (I.e. it works when the server sees that as a request for a resource and can figure out the path on it's file system based on the root of the site). I've considered trying to rewrite those kind of links so the output HTML contains the correct link including `LinkRoot` setting but haven't ever gotten far enough along to look closer.

**Reply by daveaglick** [reply]
> Is it normal to have every markdown link have a warning

It's definitely not intended. In fact, the link validation analyzer was designed to support the `LinkRoot` setting. I wonder if the way `.PrependLinkRoot(true)` is changing the links is somehow interfering with the analyzer (I tested with relative links and `LinkRoot`, there's even [unit tests](https://github.com/statiqdev/Statiq.Web/blob/main/tests/Statiq.Web.Tests/Analyzers/Html/Analyzers/ValidateRelativeLinksFixture.cs#L66), but not `PrependLinkRoot()`). Let me dig a little deeper on this one and get back to you.

**Reply by daveaglick** [reply]
> I'm also running into this issue in the markdown files now...I don't see how the workaround will work with markdown

Yeah, good call since `GetLink()` is being called by the module and not directly. I'm not even sure the idea of adding an overload that takes a fragment would work in that case. The `GetLink()` will need to accept a fragment as part of the link and treat it as such without any extra hinting. I'll follow up with some more thoughts in https://github.com/statiqdev/Statiq.Framework/issues/170.
